### ISSUE TITLE
download_helper win7 ie8

### ISSUE THREAD

**Issue by iMarkCampbell** [bcit-ci/CodeIgniter#3778]
Seems to be an issue with the download helper in CI3 - specifically on Windows7 ie8 - I assume it is to do with the response headers, as the helper from CI2.1 works fine.

**Comment by kakysha** [comment]
The best issue description - 'seems to be an issue'

**Comment by ZombieProtectionAgency** [comment]
@iMarkCampbell what kakysha said, do you have any more information about the issue? Like what it is doing, what you expect it to do, any console logs, screenshots, code usage snippets?

**Comment by iMarkCampbell** [comment]
Previously which worked fine, the headers 
if MSIE included: 
header('Pragma: public'); 
else 
header('Pragma: no-cache');

Currently all headers have: header('Pragma: no-cache'); 

Which cause the download to fail in ie8- with the error message:  "Internet Explorer was not able to open this Internet site. The requested site is either unavailable or cannot be found. Please try again later"

see line 150. - I think if MSIE pragma:public would sort this issue.

**Comment by ZombieProtectionAgency** [comment]
It looks like you are probably using CodeIgniter 3.0 then. That method used to include a check for IE, https://github.com/bcit-ci/CodeIgniter/blob/2.2-stable/system/helpers/download_helper.php#L82-L100

Also, just in case it helpes, It looks like it got shifted around a lot and eventually removed when https://github.com/bcit-ci/CodeIgniter/issues/1146 was fixed.

**Comment by iMarkCampbell** [comment]
Don't know much about headers either - but using pragma:public does the trick. Perhaps someone with a better understand of headers will step in..

**Comment by mwhitneysdsu** [comment]
1) Is the link in question using SSL (https)?
2) Is the file in a format which IE typically uses a plug-in to open, such as PDF, one of the Office formats, etc.?
3) Does the link work if you right-click, save as... instead of just following the link?
4) Does the same thing occur in newer versions of IE?

One of the reasons this may happen is that IE (the information I can find refers to older versions, I don't know if this is still the case) treats itself as a proxy for any plugin used to view a file, especially when SSL is enabled for the connection.

In theory, if both cache-control and pragma are set, a client which understands cache-control will ignore the pragma. Further, pragma is a _request_ header, so the behavior when used as a _response_ header is not defined by the standard, anyway.

In practice, IE may not even understand 'pragma: public', it just ignores it because it's not 'pragma: no-cache'.

**Comment by iMarkCampbell** [comment]
@mwhitneysdsu 
1) using http
2) Yes, it happened for pdf docx etc.
3) Not a link. Document is pulled from a database and base64_decode() is passed  to the force_download with the document name...
4) NO. Works fine.

As mentioned previously, currently using the helper from CI 2.1 - with no issues. Hope this helps someone.

**Comment by ZombieProtectionAgency** [comment]
> 3) Not a link. Document is pulled from a database and base64_decode() is passed to the force_download with the document name...

They meant instead of navigating to the page make it a link, right click, and save as. If that works it could be something to do with IEs file proxy/plugin.

**Comment by punjabik** [comment]
I can confirm this. After upgrade to CI 3 have some users problem with downloading files.
We use force_download function in download helper.

I have detailed info only from one user. He use:
OS: WinXP
Browser: IE 8.0
HTTP_USER_AGENT: Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0; GTB7.5; InfoPath.2; .NET CLR 2.0.50727; .NET CLR 3.0.4506.2152; .NET CLR 3.5.30729)

**Comment by punjabik** [comment]
I wrote patch based on iMarkCampbell conclusions and IEInternals blog: http://blogs.msdn.com/b/ieinternals/archive/2009/10/02/internet-explorer-cannot-download-over-https-when-no-cache.aspx

Firstly i just try switch words no-cache, no-store, but not solved this. Main problem for IE8 is header('Pragma: no-cache'), so I move this into condition and for IE add Pragma: public.

I have confirmed, that this works.
